crates/ruff_linter/src/external/ast/registry.rs

```diff
@@ -0,0 +1,163 @@
+            return Err(ExternalLinterError::DuplicateLinter { id: linter.id });
+        }
+
+        let linter_index = self.linters.len();
+        for (rule_index, rule) in linter.rules.iter().enumerate() {
+            let code = rule.code.as_str().to_string();
+            if self.index_by_code.contains_key(&code) {
+                return Err(ExternalLinterError::DuplicateRule {
+                    linter: linter.id.clone(),
+                    code,
+                });
+            }
+            self.index_by_code
+                .insert(code, RuleLocator::new(linter_index, rule_index));
+
+            if !linter.enabled {
+                continue;
+            }
+
+            for target in &rule.targets {
+                match target {
+                    AstTarget::Stmt(kind) => self
+                        .stmt_index
+                        .entry(*kind)
+                        .or_default()
+                        .push(RuleLocator::new(linter_index, rule_index)),
+                    AstTarget::Expr(kind) => self
+                        .expr_index
+                        .entry(*kind)
+                        .or_default()
+                        .push(RuleLocator::new(linter_index, rule_index)),
+                }
+            }
+        }
+        self.linters.push(linter);
+        Ok(())
+    }
+
+    pub fn get_rule(&self, locator: RuleLocator) -> Option<&ExternalAstRule> {
+        self.linters
+            .get(locator.linter_index)
+            .and_then(|linter| linter.rules.get(locator.rule_index))
+    }
+
+    pub fn get_linter(&self, locator: RuleLocator) -> Option<&ExternalAstLinter> {
+        self.linters.get(locator.linter_index)
+    }
+
+    pub fn find_rule_by_code(
+        &self,
+        code: &str,
+    ) -> Option<(RuleLocator, &ExternalAstRule, &ExternalAstLinter)> {
+        let locator = *self.index_by_code.get(code)?;
+        let linter = self.linters.get(locator.linter_index)?;
+        if !linter.enabled {
+            return None;
+        }
+        let rule = linter.rules.get(locator.rule_index)?;
+        Some((locator, rule, linter))
+    }
+
+    pub fn rules_for_stmt(&self, kind: StmtKind) -> impl Iterator<Item = RuleLocator> + '_ {
+        self.stmt_index.get(&kind).into_iter().flatten().copied()
+    }
+
+    pub fn rules_for_expr(&self, kind: ExprKind) -> impl Iterator<Item = RuleLocator> + '_ {
+        self.expr_index.get(&kind).into_iter().flatten().copied()
+    }
+
+    pub fn rule_entry(
+        &self,
+        locator: RuleLocator,
+    ) -> Option<(&ExternalAstRule, &ExternalAstLinter)> {
+        let linter = self.linters.get(locator.linter_index)?;
+        let rule = linter.rules.get(locator.rule_index)?;
+        Some((rule, linter))
+    }
+}
+
+impl ruff_cache::CacheKey for ExternalLintRegistry {
+    fn cache_key(&self, key: &mut ruff_cache::CacheKeyHasher) {
+        key.write_usize(self.linters.len());
+        for linter in &self.linters {
```

Comment:
This might be more tricky. Just hashing the linter metadata isn't enough. We'd also need to hash the linter implementation, e.g. by hasing the linter file content and all its dependencies. 


I suspect that we want the linter to provide its own hash during initialization.